### PR TITLE
New Recipes + Unit Tests + Bug Fix | 3 September 2024

### LINKED ISSUE #1: Change Package name
- [ ] Change package name for `com.azure.core.credential` to `io.clientcore.core.credential`

### LINKED ISSUE #7: Recipe for converting BinaryData
```[tasklist]
### Tasks
- [ ] toObject in axure-core-v2 takes an implementation of ParameterizedType instead of TypeReference (v1)
```

### PR DESCRIPTION
**Pull request by AliSoltanianFJ**
### New Recipes:

- Added recipes to convert `BinaryData`
    - Added a recipe to change instances of `TypeReference` in `toObject()` to `ParameterizedType`
    - Added a recipe to convert `com.azure.core.util.BinaryData` to `io.clientcore.core.util.binarydata.BinaryData`

- Added recipe to convert `com.azure.core.util.Configuration` to `io.clientcore.core.util.configuration.Configuration`

- Added recipe to convert `com.azure.core.util.logging.ClientLogger` to `io.clientcore.core.util.ClientLogger`

- Added recipes to convert `ClientAuthenticationException`, `HttpResponseException`, `ResourceModifiedException` and `ResourceNotFoundException`

- Added recipe to convert `com.azure.core.credential` package name to `io.clientcore.core.credential`

### Other Changes and Bug Fixes:
- Fixed a bug in the `RetryOptionsConstructorRecipe` that would cause it to crash if it encountered an initializer without `new` keyword
- Added unit tests for all of the new recipes

### Resolved Issues:
resolves #7
resolves #1 

### Discarded Changes:
- ~Added recipes to convert `@ UnexpectedResponseExceptionType` annotations and its attributes~
    - ~Added recipe to change annotation attribute names~
    - ~Added recipe to convert `com.azure.core.annotation.UnexpectedResponseExceptionType` to `io.clientcore.core.http.annotation.UnexpectedResponseExceptionDetail`~
    - ~Added recipe to convert given attribute values to azure core v2 naming convention. For example, `ClientAuthenticationException.class` is changed to `CLIENT_AUTHENTICATION`~

- ~Added recipes to convert `@ HostParam`, `@ HeaderParam`, `@ BodyParam`, and `@ QueryParam` annotations~

-~Added recipes to convert `@ ServiceClient`, `@ ServiceClientBuilder`, `@ ServiceInterface`, and `@ ServiceMethod` annotations~

- ~Added recipes to convert `@ Generated`, `@ Immutable`, and `@ ReturnType` annotations~

### CODE REVIEW

File: rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml [hunk 1]
```diff
@@ -10,7 +10,30 @@ recipeList:
...
   #    replace: 'requires $1com.azure.core.v2'
   #    filePattern: '**/module-info.java'
+
+  # ----------------------------------
+  #
+  # Java-Based LST Replacement Recipes
   #
+  # ----------------------------------
+
+  # Recipes that converts the parameters and type of the @UnexpectedResponseExceptionType annotation
```
Comment (srnagar) [inline rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml]: We don't need to convert annotations. Users don't need to use the annotations directly. Annotations are only needed by the Azure libraries. So, we don't need to update them using OpenRewrite recipes.

File: rewrite-java-core/src/test/java/BinaryDataTest.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Override
+    public void defaults(RecipeSpec spec) {
+        spec.recipeFromResource("/META-INF/rewrite/rewrite.yml",
+                "com.azure.rewrite.java.core.MigrateAzureCoreSamplesToAzureCoreV2");
+    }
+
+    /* Test to make sure BinaryData type and import is changed */
+    @Test
+    public void testClientLoggerTypeAndImportChanged() {
```
Comment (srnagar) [inline rewrite-java-core/src/test/java/BinaryDataTest.java]: We should also add tests that use TypeReference in BinaryData.

File: rewrite-sample/azure-ai-translation-text-v2/libs/com/azure/azure-core-v2/1.0.0-beta.1/_remote.repositories [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#NOTE: This is a Maven Resolver internal implementation file, its format can be changed without prior notice.
```
Comment (srnagar) [inline rewrite-sample/azure-ai-translation-text-v2/libs/com/azure/azure-core-v2/1.0.0-beta.1/_remote.repositories]: We should not check in this `libs` directory.

File: rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml (hunk 1 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml]: Removed recipes that convert annotations and their associated unit tests in commit [5447e27](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/5447e27b96a32983846088fe796effd5c4256353)

File: rewrite-java-core/src/test/java/BinaryDataTest.java (hunk 2 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/test/java/BinaryDataTest.java]: Added unit test that tests BinaryData change with TypeReference in commit [5447e27](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/5447e27b96a32983846088fe796effd5c4256353)

File: rewrite-sample/azure-ai-translation-text-v2/libs/com/azure/azure-core-v2/1.0.0-beta.1/_remote.repositories (hunk 3 above)
Comment (AliSoltanianFJ) [reply rewrite-sample/azure-ai-translation-text-v2/libs/com/azure/azure-core-v2/1.0.0-beta.1/_remote.repositories]: Removed libs directory in commit [5447e27](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/5447e27b96a32983846088fe796effd5c4256353)

File: .github/workflows/maven.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      - uses: actions/setup-java@v4
+        with:
+          distribution: 'corretto'
+          java-version: '21'
+          cache: 'maven'
+
+      - name: Clean install
+        run: mvn clean install
+      
+      - name: Execute Unit Tests
```
Comment (srnagar) [inline .github/workflows/maven.yml]: Do we need this extra step? `mvn clean install` in the previous step also runs the unit tests.

File: rewrite-java-core/src/main/java/com/azure/recipes/v2recipes/TypeReferenceRecipe.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+public class TypeReferenceRecipe extends Recipe {
+
+    @Override
+    public @NotNull String getDisplayName() {
+        return "Convert TypeReference to ParameterizedType and remove imports";
+    }
+
+    @Override
+    public @NotNull String getDescription() {
+        return "This recipe converts TypeReference<List<T>> to ParameterizedType and removes the import statement for TypeReference.";
```
Comment (srnagar) [inline rewrite-java-core/src/main/java/com/azure/recipes/v2recipes/TypeReferenceRecipe.java]: This should replace not just `List<T>`, it should replace any generic types like `Map<T,V>`, `Set<T>`, `ArrayList<T>` and even some custom generic types like `Foo<T>`. So, this recipe has to be handle all generic types.

Also, `TypeReference` can be used on non-generic classes too and those have to be converted to `Type` as well.

File: rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml [hunk 6]
```diff
@@ -21,4 +27,58 @@ recipeList:
...
+      oldFullyQualifiedTypeName: com.azure.core.util.logging.ClientLogger
+      newFullyQualifiedTypeName: io.clientcore.core.util.ClientLogger
+
+  # Recipes to migrate implementations of com.azure.core.client.traits.HttpTrait
+  # to io.clientcore.core.models.traits.HttpTrait
+  # Add http prefix to renamed methods
+  # TODO: Filter for implementing classes
+  - org.openrewrite.java.ChangeMethodName:
+      methodPattern: com.azure.core.client.traits.HttpTrait retryOptions(..)
+      newMethodName: httpRetryOptions
```
Comment (srnagar) [inline rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml]: Doesn't this need to update the package name as well?

File: rewrite-java-core/src/test/java/BinaryDataTest.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
+import com.azure.recipes.v2recipes.TypeReferenceRecipe;
```
Comment (srnagar) [inline rewrite-java-core/src/test/java/BinaryDataTest.java]: These tests should also be in the root package  - `com.azure.recipes.core.v2`. (See my comment above about changing the root package name)

File: rewrite-java-core/src/main/java/com/azure/recipes/v2recipes/TypeReferenceRecipe.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
+package com.azure.recipes.v2recipes;
```
Comment (srnagar) [inline rewrite-java-core/src/main/java/com/azure/recipes/v2recipes/TypeReferenceRecipe.java]: Instead of `v2recipes` we should call this `core.v2` i.e `com.azure.recipes.core.v2`

File: rewrite-java-core/src/test/java/BinaryDataTest.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        spec.recipe(new TypeReferenceRecipe());
+    }
+
+    /* Test to make sure BinaryData type and import is changed */
+    @Test
+    public void testBinaryDataTypeAndImportChanged() {
+        @Language("java") String before = "import com.azure.core.util.BinaryData;";
+        before += "\npublic class Testing {";
+        before += "\n  public Testing(){";
+        before += "\n    com.azure.core.util.BinaryData b = BinaryData.fromObject(null);";
```
Comment (srnagar) [inline rewrite-java-core/src/test/java/BinaryDataTest.java]: We have an import above. So, do we need to use fully qualified name here?

File: rewrite-java-core/src/test/java/CredentialTest.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+     * This test method is used to make sure that the package com.azure.core.credential is changed
+     */
+    @Test
+    void testCredentialPackageNameChange() {
+        @Language("java") String before = "import com.azure.core.credential.KeyCredential;";
+        before += "\npublic class Testing {";
+        before += "\n  public Testing(){";
+        before += "\n    com.azure.core.credential.KeyCredential kc = new KeyCredential(\"<api-key>\");";
```
Comment (srnagar) [inline rewrite-java-core/src/test/java/CredentialTest.java]: Same here - we should not use fully qualified names when there's an import. Also, we should add another test where there's not import and the fully qualified name is used.

File: rewrite-java-core/src/test/java/ExceptionTypesTest.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        before += "\npublic class Testing {";
+        before += "\n  public Testing(){";
+        before += "\n    com.azure.core.exception.ClientAuthenticationException e = new ClientAuthenticationException(null,null);";
+        before += "\n  }";
+        before += "\n}";
+
+        @Language("java") String after = "import com.azure.core.v2.exception.ClientAuthenticationException;";
+        after += "\n\npublic class Testing {";
+        after += "\n  public Testing(){";
+        after += "\n    com.azure.core.v2.exception.ClientAuthenticationException e = new ClientAuthenticationException(null,null);";
```
Comment (srnagar) [inline rewrite-java-core/src/test/java/ExceptionTypesTest.java]: Users generally don't new up exception types. They just handle these exceptions in a try-catch block. So, we should have tests for converting the exceptions in `catch` block.

File: rewrite-java-core/src/test/java/HttpTraitTest.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    void declarativeRenameMethodsSuccessful() {
+        @Language("java") String before = "import com.azure.core.client.traits.HttpTrait;\n" +
+                "import com.azure.core.http.HttpClient;\n" +
+                "import com.azure.core.http.HttpPipeline;\n" +
+                "import com.azure.core.http.policy.HttpLogOptions;\n" +
+                "import com.azure.core.http.policy.HttpPipelinePolicy;\n" +
+                "import com.azure.core.http.policy.RetryOptions;\n" +
+                "import com.azure.core.util.ClientOptions;\n" +
+                "\n" +
+                "public class BlankHttpTrait implements HttpTrait<BlankHttpTrait> {\n" +
```
Comment (srnagar) [inline rewrite-java-core/src/test/java/HttpTraitTest.java]: Users generally don't implement traits. We should test how the builder code is transformed by the recipes. For e.g. `clientBuilder.addPolicy(new HttpPipelinePolicy(){...})` is transformed to `clientBuilder.addHttpPipelinePolicy(new HttpPipelinePolicy(){...})`

File: rewrite-java-core/src/test/java/RetryOptionsConstructorTest.java [hunk 13]
```diff
@@ -56,16 +56,20 @@ void testChangeRetryOptionsTypeNoArgInit() {
         @Language("java") String before = "import com.azure.core.http.policy.RetryOptions;import java.time.Duration;import com.azure.core.http.policy.FixedDelayOptions;";
         before += "\npublic class Testing {";
         before += "\n  FixedDelayOptions f = new FixedDelayOptions(3, Duration.ofMillis(50));";
+        before += "\n  String test = \"testing\";";
```
Comment (srnagar) [inline rewrite-java-core/src/test/java/RetryOptionsConstructorTest.java]: Why is this needed?

File: rewrite-java-core/src/test/java/TypeReferenceTest.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+ * @author Ali Soltanian Fard Jahromi
+ */
+public class TypeReferenceTest implements RewriteTest {
+
+    /**
+     * This method sets which recipe should be used for testing
+     * @param spec stores settings for testing environment; e.g. which recipes to use for testing
+     */
+    @Override
+    public void defaults(RecipeSpec spec) {
```
Comment (srnagar) [inline rewrite-java-core/src/test/java/TypeReferenceTest.java]: We should import all recipes to ensure we test how all recipes interact with each other. There might be some overlap that might cause some overwriting to happen and result in unexpected transformation. So, I would suggest having a base test class where all recipes are added in RecipeSpec and all other tests inheriting from it.

File: .github/workflows/maven.yml (hunk 4 above)
Comment (AliSoltanianFJ) [reply .github/workflows/maven.yml]: Removed in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

File: rewrite-java-core/src/main/java/com/azure/recipes/v2recipes/TypeReferenceRecipe.java (hunk 5 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/main/java/com/azure/recipes/v2recipes/TypeReferenceRecipe.java]: Added functionality to replace other generic and non-generic types for TypeReference in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

File: rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml (hunk 6 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml]: The HttpTrait recipes are currently unfinished and weren't supposed to be included in this PR (were included by mistake). Removed in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

File: rewrite-java-core/src/test/java/BinaryDataTest.java (hunk 7 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/test/java/BinaryDataTest.java]: Changed package name in commit [8e6d539](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/8e6d539b5b0576d7b7e17a656beac7f15941b3c8)

File: rewrite-java-core/src/main/java/com/azure/recipes/v2recipes/TypeReferenceRecipe.java (hunk 8 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/main/java/com/azure/recipes/v2recipes/TypeReferenceRecipe.java]: Changed in commit [8e6d539](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/8e6d539b5b0576d7b7e17a656beac7f15941b3c8)

File: rewrite-java-core/src/test/java/BinaryDataTest.java (hunk 9 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/test/java/BinaryDataTest.java]: Fixed in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

File: rewrite-java-core/src/test/java/CredentialTest.java (hunk 10 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/test/java/CredentialTest.java]: Done in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

File: rewrite-java-core/src/test/java/ExceptionTypesTest.java (hunk 11 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/test/java/ExceptionTypesTest.java]: Modified exception tests to test out exceptions in catch blocks in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

File: rewrite-java-core/src/test/java/HttpTraitTest.java (hunk 12 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/test/java/HttpTraitTest.java]: The HttpTrait recipes are currently unfinished and weren't supposed to be included in this PR (were included by mistake). Removed in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

File: rewrite-java-core/src/test/java/RetryOptionsConstructorTest.java (hunk 13 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/test/java/RetryOptionsConstructorTest.java]: Removed in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

File: rewrite-java-core/src/test/java/TypeReferenceTest.java (hunk 14 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/test/java/TypeReferenceTest.java]: Ensured all testing classes only use `rewrite.yml`. Moved all imperative (Java-based) recipes into `rewrite.yml`.

See changes in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

File: rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml (hunk 6 above)
Comment (stoatus-quo) [reply rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml]: Package name will change for all HttpTrait methods when HttpTrait is changed. This is safer to change as the last thing.

File: .gitignore [hunk 15]
```diff
@@ -410,7 +410,6 @@ FodyWeavers.xsd
 .mtj.tmp/
 
 # Package Files #
-*.jar
```
Comment (srnagar) [inline .gitignore]: nit: we should not check-in jar files

**Review by srnagar — APPROVED** [review]
Looks good to me

File: .gitignore (hunk 15 above)
Comment (AliSoltanianFJ) [reply .gitignore]: Addressed in commit [a29f16e](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/a29f16e6f730a6332908c520c0bd5244ffe4406c), should be ready to merge

File: rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java [hunk 16]
```diff
@@ -1,4 +1,4 @@
-package com.azure.recipes.v2recipes;
+package com.azure.recipes.v2;
 
 import org.jetbrains.annotations.NotNull;
 import org.openrewrite.*;
```
Comment (samvaity) [inline rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java]: nit: we should avoid using star imports.

File: rewrite-java-core/src/test/java/com/azure/recipes/v2/TypeReferenceTest.java [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+     * changed to ParameterizedType when using Map generic type
+     */
+    @Test
+    void testTypeReferenceVariableDeclarationChangeMap() {
+        @Language("java") String before = "";
+        before += "\nimport java.lang.reflect.ParameterizedType;";
+        before += "\nimport java.lang.reflect.Type;";
+        before += "\nimport com.azure.core.util.serializer.TypeReference;";
+        before += "\npublic class Testing {";
+        before += "\n  private static final TypeReference<java.util.Map<String,Integer>> TESTING_TYPE = new TypeReference<java.util.Map<String, Integer>>() {\n  };";
```
Comment (samvaity) [inline rewrite-java-core/src/test/java/com/azure/recipes/v2/TypeReferenceTest.java]:
```suggestion
before += "\n  private static final TypeReference<java.util.Map<String, Integer>> TESTING_TYPE = new TypeReference<java.util.Map<String, Integer>>() {\n  };";
```

File: rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java [hunk 18]
```diff
@@ -47,15 +47,30 @@ private static class ConvertTypeReferenceVisitor extends JavaIsoVisitor<Executio
                     .anyMatch(methodDeclaration -> methodDeclaration.getName().getSimpleName().equals("getRawType"));
 
             if (!alreadyTransformed && visitedNewClass.toString().contains("TypeReference")) {
-                // Extract type from List type in TypeReference declaration
-                Pattern pattern = Pattern.compile("List<(\\w+)>");
+                // Extract type from generic type in TypeReference declaration
+                Pattern pattern = Pattern.compile("<([\\w.]+)<([\\w.]+)(?:,\\s*([\\w.]+))?>>");
```
Comment (samvaity) [inline rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java]: Instead of parsing the string representation of the class, use the structured metadata provided by `J.NewClass`
here https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java
This will help in giving you access to its type arguments, same way without the regex.

File: rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java [hunk 19]
```diff
@@ -47,15 +47,30 @@ private static class ConvertTypeReferenceVisitor extends JavaIsoVisitor<Executio
                     .anyMatch(methodDeclaration -> methodDeclaration.getName().getSimpleName().equals("getRawType"));
 
             if (!alreadyTransformed && visitedNewClass.toString().contains("TypeReference")) {
```
Comment (samvaity) [inline rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java]: Same suggestion here the inferred type can be used here instead of string comparison to `TypeReference`.

File: rewrite-java-core/src/test/java/com/azure/recipes/core/v2/TypeReferenceTest.java [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+        before += "\n}";
+
+
+        @Language("java") String after = "import java.lang.reflect.ParameterizedType;\n" +
+                "import java.lang.reflect.Type;\n" +
+                "public class Testing {\n" +
+                "  private static final java.lang.reflect.Type TESTING_TYPE = new ParameterizedType() {\n" +
+                "      @Override\n" +
+                "      public java.lang.reflect.Type getRawType() {\n" +
+                "          return java.util.List.class;\n" +
```
Comment (samvaity) [inline rewrite-java-core/src/test/java/com/azure/recipes/core/v2/TypeReferenceTest.java]: We are still using fully qualified names here and other places.

Earlier comment https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14#discussion_r1750823950

File: rewrite-java-core/src/main/java/com/azure/recipes/core/v2/TypeReferenceRecipe.java [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+            // Check if the TypeReference reference has already been transformed
+            if (visitedNewClass.getBody() == null){return visitedNewClass;}
+            boolean alreadyTransformed = visitedNewClass.getBody().getStatements().stream()
+                    .filter(statement -> statement instanceof J.MethodDeclaration)
+                    .map(J.MethodDeclaration.class::cast)
+                    .anyMatch(methodDeclaration -> methodDeclaration.getName().getSimpleName().equals("getRawType"));
+
+            if (!alreadyTransformed && visitedNewClass.toString().contains("TypeReference")) {
+                // Extract type from generic type in TypeReference declaration
+                Pattern pattern = Pattern.compile("<([\\w.]+)<([\\w.]+)(?:,\\s*([\\w.]+))?>>");
```
Comment (samvaity) [inline rewrite-java-core/src/main/java/com/azure/recipes/core/v2/TypeReferenceRecipe.java]: https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14#discussion_r1771947112

J.ParameterizedType will give you access to the arguments instead of Regex matcher here.

File: rewrite-java-core/src/main/java/com/azure/recipes/core/v2/TypeReferenceRecipe.java [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+
+        /**
+         * Method to remove TypeReference import and replace it with ParameterizedType.
+         * Also makes sure that no duplicate imports of ParameterizedType are created.
+         */
+        int count = 0;
+        @Override
+        public J.@NotNull Import visitImport(J.@NotNull Import importStmt, @NotNull ExecutionContext ctx) {
+            if (importStmt.getQualid().toString().equals("java.lang.reflect.ParameterizedType")) {
+                count ++;
```
Comment (samvaity) [inline rewrite-java-core/src/main/java/com/azure/recipes/core/v2/TypeReferenceRecipe.java]: Instead of using the count variable for tracking imports (which can be error-prone), you could keep track of the imports in a set.

File: rewrite-java-core/src/main/java/com/azure/recipes/core/v2/TypeReferenceRecipe.java [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+            J.NewClass visitedNewClass = super.visitNewClass(newClass, ctx);
+
+            // Check if the TypeReference reference has already been transformed
+            if (visitedNewClass.getBody() == null){return visitedNewClass;}
+            boolean alreadyTransformed = visitedNewClass.getBody().getStatements().stream()
+                    .filter(statement -> statement instanceof J.MethodDeclaration)
+                    .map(J.MethodDeclaration.class::cast)
+                    .anyMatch(methodDeclaration -> methodDeclaration.getName().getSimpleName().equals("getRawType"));
+
+            if (!alreadyTransformed && visitedNewClass.toString().contains("TypeReference")) {
```
Comment (samvaity) [inline rewrite-java-core/src/main/java/com/azure/recipes/core/v2/TypeReferenceRecipe.java]: Could use if (visitedNewClass.getClazz() instanceof J.ParameterizedType) {
?

File: rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java (hunk 16 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java]: Addressed in latest commits.

File: rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java (hunk 18 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java]: Addressed in latest commits.

File: rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java (hunk 19 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java]: Addressed in latest commits.

File: rewrite-java-core/src/test/java/com/azure/recipes/core/v2/TypeReferenceTest.java (hunk 20 above)
Comment (AliSoltanianFJ) [reply rewrite-java-core/src/test/java/com/azure/recipes/core/v2/TypeReferenceTest.java]: Addressed in latest commits. Fully qualified names no longer used when imports are already added.

**Comment by AliSoltanianFJ** [comment]
Changes according to feedback/reviews:
- Ensured fully qualified types are not used whenever an import is already added
- Extracted type information in TypeReferenceRecipe where possible
- Fixed a bug in TypeReference recipe that would not add the required import for `java.lang.reflect.Type`
- Removed type validation from TypeReference test due to bug in OpenRewrite test framework parser. Note that the tests still pass as the transformed code matches expected output.

**Comment by AliSoltanianFJ** [comment]
Changes according to feedback/reviews:
- Ensured fully qualified types are not used in unit testing code whenever an import is already added
- Extracted type information in TypeReferenceRecipe where possible
- Fixed a bug in TypeReference recipe that would not add the required import for `java.lang.reflect.Type`
- Fixed a bug in RetryOptionsConstructor recipe which would cause it to crash when it parsed certain Java types while searching for RetryOptions
- Removed type validation from TypeReference test due to bug in OpenRewrite test framework parser that would not allow it to parse the instantiation of TypeReference. Note that the tests still pass as the transformed code matches expected output.
- Removed usages of star (*) imports

**Review by samvaity — APPROVED** [review]
